sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/context/AuthTokenContext.java

```diff
@@ -131,47 +173,160 @@ private void updateAuthToken(String authToken, String homeAccountId) {
     }
 
     /**
-     * Clears info about the token.
+     * Gets current homeAccountId value.
+     *
+     * @return unique identifier of user.
      */
-    public synchronized void clearAuthToken() {
-        setAuthToken(null, null, null);
+    public synchronized String getHomeAccountId() {
+        List<AuthTokenHistoryEntry> history = getHistory();
+        if (history != null && history.size() > 0) {
+            return history.get(history.size() - 1).getHomeAccountId();
+        }
+        return null;
     }
 
     /**
-     * Cache auth token and account data to be used later on.
+     * Gets list of auth tokens validity info. It contains tokens and time when it was valid.
+     *
+     * @return auth token info.
+     * @see AuthTokenInfo
      */
-    public synchronized void cacheAuthToken() {
-        if (mStorage != null) {
-            updateAuthToken(mStorage.getToken(), mStorage.getHomeAccountId());
+    @NonNull
+    public synchronized List<AuthTokenInfo> getAuthTokenValidityList() {
+        List<AuthTokenHistoryEntry> history = getHistory();
+        if (history == null || history.size() == 0) {
+            return Collections.singletonList(new AuthTokenInfo());
         }
+
+        /* Return history with corrected end times. */
+        List<AuthTokenInfo> result = new ArrayList<>();
+        if (history.get(0).getAuthToken() != null) {
+            result.add(new AuthTokenInfo(null, null, history.get(0).getTime()));
+        }
+        for (int i = 0; i < history.size(); i++) {
+            AuthTokenHistoryEntry storeEntity = history.get(i);
+            String token = storeEntity.getAuthToken();
+            Date startTime = storeEntity.getTime();
+            if (token == null && i == 0) {
+                startTime = null;
+            }
+            Date endTime = storeEntity.getExpiresOn();
+            Date nextChangeTime = history.size() > i + 1 ? history.get(i + 1).getTime() : null;
+            if (nextChangeTime != null && endTime != null && nextChangeTime.before(endTime)) {
+                endTime = nextChangeTime;
+            } else if (endTime == null && nextChangeTime != null) {
+                endTime = nextChangeTime;
+            }
+            result.add(new AuthTokenInfo(token, startTime, endTime));
+        }
+        return result;
     }
 
     /**
-     * Gets token storage.
+     * Removes the token from history. Please note that only oldest token is
+     * allowed to remove. To reset current to anonymous, use
+     * {@link #setAuthToken(String, String, Date)} with <code>null</code> value instead.
      *
-     * @return token storage.
+     * @param token auth token to remove. Despite the fact that only the oldest
+     *              token can be removed, it's required to avoid removing
+     *              the wrong one on duplicated calls etc.
      */
-    public AuthTokenStorage getStorage() {
-        return mStorage;
+    public synchronized void removeToken(String token) {
```

Comment:
removeOldestTokenIfMatching or something like that? It's doing something very specific.